### PR TITLE
[ISSUE-1620] Add Azure Blob Storage (ABS) support to BentoML

### PR DESCRIPTION
**Pull request by alxmke**
## Description

Add Azure Blob Storage (ABS) support to BentoML's Yatai service.

## Motivation and Context

This change integrates ABS into BentoML's Yatai service.

## How Has This Been Tested?

I manually tested all the Azure functions in a personal sandbox repository to assure they behaved as expected. I wrote unit tests which are analogous to those already in place for s3 and gcs.

I ran tests on a fresh installation of BentoML, on my local machine, and with my personal Azure account.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature and improvements (non-breaking change which adds/improves functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Code Refactoring (internal change which is not user facing)
- [x] Documentation
- [x] Test, CI, or build

## Component(s) if applicable
- [ ] BentoService (service definition, dependency management, API input/output adapters)
- [ ] Model Artifact (model serialization, multi-framework support)
- [ ] Model Server (mico-batching, dockerisation, logging, OpenAPI, instruments)
- [x] YataiService gRPC server (model registry, cloud deployment automation)
- [ ] YataiService web server (nodejs HTTP server and web UI)
- [ ] Internal (BentoML's own configuration, logging, utility, exception handling)
- [x] BentoML CLI

## Checklist:

- [x] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
note: I've left this unchecked, because it makes a lot of changes to code sections, many which I haven't worked in, and some of which that I'd like to talk about relaxing or changing.
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [x] My change reduces project test coverage and requires unit tests to be added
- [x] I have added unit tests covering my code change
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly

### CODE REVIEW

**Comment by jjmachan** [comment]
Hey @alxmke awesome progress man! I've been working on the same here https://github.com/jjmachan/BentoML/tree/abs but its seems like you have most of it working now. But if you have anything I can take up feel free to tell

**Comment by jjmachan** [comment]
also about authentication, what do you think of adding auth with Azure AD functionality? it gives you auth options through the cli and a bunch of other places too.

**Comment by jjmachan** [comment]
Also, one other thing I wanted to get the opinion of the team is the order in which the commands are run during the bentoml push command, especially with remote repositories in yatai. The issue is that sometimes when I do a push to a remote yatai and if I get an error during the creation of the signed URL an exception is raised but an entry is still created in the remote yatai db. so when I try to do a push or do a list, that model is listed even though it might not be in the object store. 

I personally found it annoying but I was wondering what everyone thought of it.

**Comment by yubozhao** [comment]
@alxmke Thanks for the PR.  It's great to add Azure blob storage support for Yatai!

Can you fix the linter and format issue?  
You can run `./dev/format.sh` and `./dev/linter.sh` from the git repo root directory to format the code and see what are the issues and fix them

Can you also fix the unit tests issues?

File: bentoml/cli/yatai_service.py [hunk 1]
```diff
@@ -131,6 +133,13 @@ def add_yatai_service_sub_command(
         help='Specifies the GCS URL for the GCS repository type',
         envvar='BENTOML_YATAI_GCS_URL',
     )
+    @click.option(
+        '--abs-url',
+        type=click.STRING,
+        default=default_abs_url,
+        help='Specifies the ABS URL for the UBS repository type',
```
Comment (yubozhao) [inline bentoml/cli/yatai_service.py:140]:
```suggestion
        help='Specifies the Azure block storage URL for the Yatai repository type',
```

File: bentoml/utils/abs.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    For a container, the base URI includes the name of the account and the name of the container:
+        https://myaccount.blob.core.windows.net/mycontainer
+    For a blob, the base URI includes the name of the account, the name of the container, and the name of the blob:
+        https://myaccount.blob.core.windows.net/mycontainer/myblob
+    Source: https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#resource-uri-syntax
+    """
+    try:
+        return "blob.core.windows.net" in urlparse(url).netloc
+    except ValueError:
+        return False
```
Comment (yubozhao) [inline bentoml/utils/abs.py:34]:
```suggestion
        return False

```

File: bentoml/utils/abs.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+def is_abs_url(url):
+    """
+    Check if the url is a abs url
+    For the storage account, the base URI includes the name of the account only:
+        https://myaccount.blob.core.windows.net
+    For a container, the base URI includes the name of the account and the name of the container:
+        https://myaccount.blob.core.windows.net/mycontainer
+    For a blob, the base URI includes the name of the account, the name of the container, and the name of the blob:
+        https://myaccount.blob.core.windows.net/mycontainer/myblob
+    Source: https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#resource-uri-syntax
```
Comment (yubozhao) [inline bentoml/utils/abs.py:29]: 👍🏼

File: bentoml/cli/yatai_service.py (hunk 1 above)
Comment (yubozhao) [reply bentoml/cli/yatai_service.py:140]: Can you add an example URL for this?

File: bentoml/cli/yatai_service.py [hunk 4]
```diff
@@ -169,6 +183,9 @@ def yatai_service_start(
         elif repository_type == YATAI_REPOSITORY_GCS and gcs_url is None:
             logger.error("'--gcs-url' must be specified for GCS repository type")
             return
+        elif repository_type == YATAI_REPOSITORY_ABS and abs_url is NONE:
```
Comment (yubozhao) [inline bentoml/cli/yatai_service.py:186]:
```suggestion
        elif repository_type == YATAI_REPOSITORY_ABS and abs_url is None:
```

File: bentoml/yatai/repository/abs_repository.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        self.generate_blob_sas = generate_blob_sas
+        self.sas_permission_add = AccountSasPermissions(
+            create=True,
+            write=True,
+        )
+        self.sas_permission_get = AccountSasPermissions(
+            read=True,
+        )
+
+        self.uri_type = BentoUri.ABS
```
Comment (yubozhao) [inline bentoml/yatai/repository/abs_repository.py:61]: Can you make sure run `./proto/generate-docker.sh` to generate python and nodejs definitions? It doesn't work right now

File: bentoml/yatai/repository/abs_repository.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+                generate_blob_sas,
+                AccountSasPermissions,
+            )
+        except ImportError:
+            raise YataiRepositoryException(
+                '"azure-storage-blob" package is required for Azure Blob Storage.'
+                'You can install it with pip: '
+                '"pip install pip install azure-storage-blob"'
+                'Find out more at https://docs.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-python'  # noqa: E501
+            )
```
Comment (yubozhao) [inline bentoml/yatai/repository/abs_repository.py:50]: Let's include the `azure-storage-blob` package in the setup.py

File: bentoml/yatai/repository/abs_repository.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from datetime import datetime, timedelta
+import os
+from urllib.parse import urlparse
+from dependency_injector.wiring import Provide, inject
```
Comment (parano) [inline bentoml/yatai/repository/abs_repository.py:19]: We may need to change this import after PR https://github.com/bentoml/BentoML/pull/1697

File: bentoml/yatai/repository/abs_repository.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        # fmt: on
+
+        self.uri_type = BentoUri.ABS
+        self.base_url = base_url
+
+        parsed_url = urlparse(base_url)
+        self.account = parsed_url.netloc.split(".", 1)[0]
+        self.container, self.blob = parsed_url.path.split("/", 1)
+
+        connection_string = os.getenv("AZURE_STORAGE_CONNECTION_STRING")
```
Comment (yubozhao) [inline bentoml/yatai/repository/abs_repository.py:69]: We want to proactively throw error, if the connection string is missing

**Pull request closed without merging**